google/cloud/aiplatform/models.py

```diff
@@ -1371,51 +1881,189 @@ def list(
         """
-        self._sync_gca_resource()
+        self._sync_gca_resource_if_skipped()
 
-        models_to_undeploy = sorted(  # Undeploy zero traffic models first
-            self._gca_resource.traffic_split.keys(),
-            key=lambda id: self._gca_resource.traffic_split[id],
+        if len(self._gca_resource.deployed_models):
+            raise ValueError(
+                "A maximum of one model can be deployed to each private Endpoint. "
+                "Please call PrivateEndpoint.undeploy_all() before deploying another "
+                "model."
+            )
+
+        self._validate_deploy_args(
+            min_replica_count,
+            max_replica_count,
+            accelerator_type,
+            deployed_model_display_name,
+            traffic_split,
+            traffic_percentage,
+            explanation_metadata,
+            explanation_parameters,
         )
 
-        for deployed_model in models_to_undeploy:
-            self._undeploy(deployed_model_id=deployed_model, sync=sync)
+        self._deploy(
+            model=model,
+            deployed_model_display_name=deployed_model_display_name,
+            machine_type=machine_type,
+            min_replica_count=min_replica_count,
+            max_replica_count=max_replica_count,
+            accelerator_type=accelerator_type,
+            accelerator_count=accelerator_count,
+            service_account=service_account,
+            explanation_metadata=explanation_metadata,
+            explanation_parameters=explanation_parameters,
+            metadata=metadata,
+            sync=sync,
+        )
 
-        return self
+    def undeploy(
+        self,
+        deployed_model_id: str,
+        sync=True,
+    ) -> None:
+        """Undeploys a deployed model from the private Endpoint.
+
+        Example Usage:
+            my_private_endpoint.undeploy(
+                deployed_model_id="1234567891232567891"
+            )
+
+            or
+
+            my_deployed_model_id = my_private_endpoint.list_models()[0].id
+            my_private_endpoint.undeploy(
+                deployed_model_id=my_deployed_model_id
+            )
+
+        Args:
+            deployed_model_id (str):
+                Required. The ID of the DeployedModel to be undeployed from the
+                private Endpoint. Use PrivateEndpoint.list_models() to get the
+                deployed model ID.
+            sync (bool):
+                Whether to execute this method synchronously. If False, this method
+                will be executed in concurrent Future and any downstream object will
+                be immediately returned and synced when the Future has completed.
+        """
+        self._sync_gca_resource_if_skipped()
+
+        self._undeploy(
+            deployed_model_id=deployed_model_id,
+            sync=sync,
+        )
 
     def delete(self, force: bool = False, sync: bool = True) -> None:
```

Comment:
Why not rely on the super class(Endpoint) implementation?

Reply:
The `delete` with `force=True` in the super class(Endpoint) implementation calls undeploy_all() which sorts the models based on their traffic split percentage. It then goes through this traffic split dictionary (that maps model_id to traffic split) and deletes them one by one. Since private Endpoint has no traffic split, this dictionary ends up empty, doesn't delete the model, and raises the error to undeploy your model before deleting which is confusing. I think having this seperate `delete()` function could be more easy to understand instead of introducing an if statement in the Endpoint `delete` function just for private Endpoints

Reply:
SG. Thanks for the detailed explanation!